File: medium/1358-number-of-substrings-containing-all-three-characters/solution.java
/**
 * 1358. Number of Substrings Containing All Three Characters
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int numberOfSubstrings(String s) {
        int[] last = {-1, -1, -1}; // last index of a, b, c
        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            last[s.charAt(i) - 'a'] = i;
            int m = Math.min(last[0], Math.min(last[1], last[2]));
            ans += m + 1; // m = -1 until all three seen
        }
        return ans;
    }
}

File: medium/1358-number-of-substrings-containing-all-three-characters/solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        """
        1358. Number of Substrings Containing All Three Characters
        Time: O(n)
        Space: O(1)
        """
        last = [-1, -1, -1]  # last index of a, b, c
        ans = 0
        for i, ch in enumerate(s):
            last[ord(ch) - ord("a")] = i
            ans += min(last) + 1  # min(last) = -1 until all three seen
        return ans
